### PR TITLE
Example for memory-mapped Dask array

### LINKED ISSUE #7355: Add example for loading raw binary data with memory mapping to documentation
In https://docs.dask.org/en/latest/array-creation.html there are a number of examples on how to create dask arrays.

A very efficient method that is not listed yet is memory mapping. Here's an example for that, see also https://github.com/LiberTEM/LiberTEM/pull/984#issuecomment-792948584:

```python
def mmap_load_chunk(filename, shape, dtype, sl):
    data = np.memmap(filename, mode='r', shape=shape, dtype=dtype)
    result = data[sl]
    del data
    return result
    
def mmap_dask_array(filename, shape, dtype):
    chunks = []
    blocksize=5
    for offset in range(0, shape[0], blocksize):
        chunk = da.from_delayed(
            delayed(mmap_load_chunk)(
                filename,
                shape=shape,
                dtype=dtype,
                sl=slice(offset, min(offset+blocksize, shape[0]))
            ),
            shape=(min(offset+blocksize, shape[0]) - offset, ) + shape[1:],
            dtype=dtype
        )
        chunks.append(chunk)
    return da.concatenate(chunks, axis=0)
```

This method has a number of advantages:

* Zero copy, very fast data access from the file system cache
* Low memory consumption
* Lazy loading of the data, completely delegated to the operating system

This method is particularly efficient if only small subsets of the Dask array are accessed at a time since there is no overhead from allocating large chunks. Furthermore, this method is pretty insensitive to the chunking scheme for the same reason.

Technically one could also use `da.from_array()` on a `numpy.memmap()` object. However, that has a number of disadvantages compared to the `from_delayed` method:

* It does not work with `distributed` if the `Client` sees a different filesystem than the workers.
* It may share the memory map between workers or pickle it and send it over the network, which could be system-dependent and just introduces unnecessary potential for errors or undefined behavior. The example above creates the memory map on the system/worker where the data is actually needed, which should be clean in any scenario.

As a side note, the same consideration holds for the HDF5 example in the documentation. It might be cleaner to wrap the file opening in `dask.delayed` to make sure that the `h5py.File` object doesn't have to be pickled or shared. The disadvantage here would be that the HDF5 chunk cache is never shared. Since the chunk cache introduces a hidden state to a `h5py.File` object, sharing it might be a hidden case of this one: https://docs.dask.org/en/latest/delayed-best-practices.html#avoid-global-state In any case, accessing a dataset is idempotent independent of the state of the chunk cache, so it should at least always return the desired result.

If you are interested, I could clean up the mmap case a bit, write it up for https://docs.dask.org/en/latest/array-creation.html and/or add a `da.from_raw()` method. I could also adapt the HDF5 case to only open the file in the worker processes/threads. I'm just asking beforehand since I want to make sure this would be appreciated before investing work into it.

### PR DESCRIPTION
**Pull request by uellue**
Closes #7355

- [x] Closes #7355
- [x] Tests added / passed
        * Documentation builds and example runs in notebook. No https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html to run the examples, I guess?
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: docs/source/array-creation.rst [hunk 1]
```diff
@@ -237,6 +237,114 @@ chunk shape:
 These interactions work not just for NumPy arrays but for any object that has
 shape and dtype attributes and implements NumPy slicing syntax.
 
+Memory mapping
+--------------
+
+Memory mapping can be a highly effective method to access raw binary data since
+it has nearly zero overhead if the data is already in the file system cache. For
+a threaded scheduler, creating a Dask array from a raw binary file can be as simple as
```
Comment (jakirkham) [inline docs/source/array-creation.rst:245]:
```suggestion
the threaded scheduler, creating a Dask array from a raw binary file can be as simple as
```

**Review by jakirkham — COMMENTED** [review]
Thanks Dieter! 😄

Had a small suggestion below

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@jsignell @martindurant, do either of you have thoughts here? 🙂

**Comment by martindurant** [comment]
+1

**Comment by jsignell** [comment]
This looks good to me! Thanks @uellue for opening this :)

**Pull request merged**